File: tests/wasm/test_elastic_volume_weighted_moving_average.js
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;

function sampleSeries(length = 96) {
    const prices = new Float64Array(length);
    const volumes = new Float64Array(length);
    for (let i = 0; i < length; i++) {
        prices[i] = 100.0 + i * 0.23 + Math.sin(i * 0.17);
        volumes[i] = 1000.0 + (i % 9) * 57.0 + i * 2.0;
    }
    return { prices, volumes };
}

function naiveEvwma(prices, volumes, length = 30, absoluteVolumeMillions = 134.0, useVolumeSum = false) {
    const out = new Float64Array(prices.length);
    out.fill(NaN);
    let first = -1;
    for (let i = 0; i < prices.length; i++) {
        if (Number.isFinite(prices[i]) && Number.isFinite(volumes[i])) {
            first = i;
            break;
        }
    }
    if (first < 0) {
        return out;
    }

    const ring = new Float64Array(length);
    const absoluteVolume = absoluteVolumeMillions * 1_000_000.0;
    let rollingSum = 0.0;
    let count = 0;
    let head = 0;
    let prev = NaN;

    for (let i = first; i < prices.length; i++) {
        const price = prices[i];
        const volume = volumes[i];
        if (useVolumeSum) {
            const volumeValue = Number.isFinite(volume) ? volume : 0.0;
            if (count < length) {
                ring[count] = volumeValue;
                rollingSum += volumeValue;
                count += 1;
            } else {
                rollingSum += volumeValue - ring[head];
                ring[head] = volumeValue;
                head = (head + 1) % length;
            }
        }
        const volumePeriod = useVolumeSum ? rollingSum : absoluteVolume;
        if (!Number.isFinite(price) || !Number.isFinite(volume) || !Number.isFinite(volumePeriod) || volumePeriod === 0.0) {
            prev = NaN;
            continue;
        }
        const base = Number.isFinite(prev) ? prev : price;
        const value = ((volumePeriod - volume) * base + volume * price) / volumePeriod;
        out[i] = value;
        prev = value;
    }
    return out;
}

function assertSeriesClose(actual, expected, message) {
    assert.strictEqual(actual.length, expected.length, `${message}: length mismatch`);
    for (let i = 0; i < actual.length; i++) {
        const a = actual[i];
        const e = expected[i];
        if (Number.isNaN(a) && Number.isNaN(e)) {
            continue;
        }
        assert.ok(Math.abs(a - e) <= 1e-10, `${message}: mismatch at ${i}: ${a} vs ${e}`);
    }
}

function getMemory(mod) {
    if (mod.memory && mod.memory.buffer) {
        return mod.memory;
    }
    if (typeof mod.__wbindgen_memory === 'function') {
        const memory = mod.__wbindgen_memory();
        if (memory && memory.buffer) {
            return memory;
        }
    }
    if (mod.__wasm && mod.__wasm.memory && mod.__wasm.memory.buffer) {
        return mod.__wasm.memory;
    }
    return null;
}

function getFloat64View(ptr, len) {
    const memory = getMemory(wasm);
    assert(memory && memory.buffer, 'WASM memory export unavailable');
    return new Float64Array(memory.buffer, ptr, len);
}

test.before(async () => {
    const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
    const importPath = process.platform === 'win32'
        ? 'file:///' + wasmPath.replace(/\\/g, '/')
        : wasmPath;
    wasm = await import(importPath);
});

test('EVWMA matches naive absolute mode', () => {
    const { prices, volumes } = sampleSeries();
    const actual = wasm.elastic_volume_weighted_moving_average(prices, volumes, undefined, undefined, undefined);
    const expected = naiveEvwma(prices, volumes);
    assertSeriesClose(actual, expected, 'absolute mode');
});

test('EVWMA matches naive volume-sum mode', () => {
    const { prices, volumes } = sampleSeries();
    const actual = wasm.elastic_volume_weighted_moving_average(prices, volumes, 7, 200.0, true);
    const expected = naiveEvwma(prices, volumes, 7, 200.0, true);
    assertSeriesClose(actual, expected, 'volume-sum mode');
});

test('EVWMA batch matches single', () => {
    const { prices, volumes } = sampleSeries(84);
    const batch = wasm.elastic_volume_weighted_moving_average_batch(prices, volumes, {
        length_range: [5, 9, 2],
        absolute_volume_millions: 180.0,
        use_volume_sum: true,
    });
    assert.strictEqual(batch.rows, 3);
    assert.strictEqual(batch.cols, prices.length);
    for (let row = 0; row < batch.rows; row++) {
        const start = row * batch.cols;
        const end = start + batch.cols;
        const single = wasm.elastic_volume_weighted_moving_average(
            prices,
            volumes,
            batch.combos[row].length,
            180.0,
            true,
        );
        assertSeriesClose(batch.values.slice(start, end), single, `batch row ${row}`);
    }
});

test('EVWMA zero-copy paths work', () => {
    const { prices, volumes } = sampleSeries(48);
    const len = prices.length;
    const batchRows = 3;
    const inPricePtr = wasm.elastic_volume_weighted_moving_average_alloc(len);
    const inVolumePtr = wasm.elastic_volume_weighted_moving_average_alloc(len);
    const outPtr = wasm.elastic_volume_weighted_moving_average_alloc(batchRows * len);
    try {
        getFloat64View(inPricePtr, len).set(prices);
        getFloat64View(inVolumePtr, len).set(volumes);
        wasm.elastic_volume_weighted_moving_average_into(
            inPricePtr,
            inVolumePtr,
            outPtr,
            len,
            7,
            200.0,
            true,
        );
        const actual = Array.from(getFloat64View(outPtr, len));
        const expected = wasm.elastic_volume_weighted_moving_average(prices, volumes, 7, 200.0, true);
        assertSeriesClose(actual, expected, 'zero-copy single');

        const config = {
            length_range: [5, 9, 2],
            absolute_volume_millions: 180.0,
            use_volume_sum: true,
        };
        const rows = wasm.elastic_volume_weighted_moving_average_batch_into(
            inPricePtr,
            inVolumePtr,
            outPtr,
            len,
            config,
        );
        assert.strictEqual(rows, batchRows);
        const batch = wasm.elastic_volume_weighted_moving_average_batch(prices, volumes, config);
        assertSeriesClose(Array.from(getFloat64View(outPtr, rows * len)), batch.values, 'zero-copy batch');
    } finally {
        wasm.elastic_volume_weighted_moving_average_free(inPricePtr, len);
        wasm.elastic_volume_weighted_moving_average_free(inVolumePtr, len);
        wasm.elastic_volume_weighted_moving_average_free(outPtr, batchRows * len);
    }
});

test('EVWMA stream and reset work', () => {
    const { prices, volumes } = sampleSeries(80);
    const stream = new wasm.ElasticVolumeWeightedMovingAverageStreamWasm(9, 220.0, true);
    const batch = wasm.elastic_volume_weighted_moving_average(prices, volumes, 9, 220.0, true);
    const streamed = [];
    for (let i = 0; i < prices.length; i++) {
        const value = stream.update(prices[i], volumes[i]);
        streamed.push(value ?? NaN);
    }
    assertSeriesClose(streamed, batch, 'stream');
    stream.reset();
    const streamedAgain = [];
    for (let i = 0; i < prices.length; i++) {
        const value = stream.update(prices[i], volumes[i]);
        streamedAgain.push(value ?? NaN);
    }
    assertSeriesClose(streamedAgain, batch, 'stream reset');
});

test('EVWMA errors', () => {
    const { prices, volumes } = sampleSeries(16);
    assert.throws(
        () => wasm.elastic_volume_weighted_moving_average([], [], undefined, undefined, undefined),
        /input data slice is empty|Input data slice is empty/
    );
    assert.throws(
        () => wasm.elastic_volume_weighted_moving_average(prices, volumes.subarray(0, 8), undefined, undefined, undefined),
        /length mismatch/
    );
    assert.throws(
        () => wasm.elastic_volume_weighted_moving_average(prices, volumes, 0, undefined, true),
        /invalid length|Invalid length/
    );
    assert.throws(
        () => wasm.elastic_volume_weighted_moving_average(prices, volumes, undefined, 0.0, false),
        /invalid absolute_volume_millions|Invalid absolute_volume_millions/
    );
});

File: tests/python/test_elastic_volume_weighted_moving_average.py
import numpy as np
import pytest

try:
    import vector_ta as ta_indicators
except ImportError:
    try:
        import vector_ta as ta_indicators
    except ImportError:
        pytest.skip("Python module not built.", allow_module_level=True)


def naive_evwma(prices, volumes, length=30, absolute_volume_millions=134.0, use_volume_sum=False):
    prices = np.asarray(prices, dtype=float)
    volumes = np.asarray(volumes, dtype=float)
    out = np.full(prices.shape, np.nan, dtype=float)
    valid = np.flatnonzero(np.isfinite(prices) & np.isfinite(volumes))
    if valid.size == 0:
        return out

    first = int(valid[0])
    prev = np.nan
    absolute_volume = absolute_volume_millions * 1_000_000.0
    ring = np.zeros(length, dtype=float)
    rolling_sum = 0.0
    count = 0
    head = 0

    for idx in range(first, len(prices)):
        price = prices[idx]
        volume = volumes[idx]
        if use_volume_sum:
            volume_value = volume if np.isfinite(volume) else 0.0
            if count < length:
                ring[count] = volume_value
                rolling_sum += volume_value
                count += 1
            else:
                rolling_sum += volume_value - ring[head]
                ring[head] = volume_value
                head = (head + 1) % length
        volume_period = rolling_sum if use_volume_sum else absolute_volume
        if (
            not np.isfinite(price)
            or not np.isfinite(volume)
            or not np.isfinite(volume_period)
            or volume_period == 0.0
        ):
            prev = np.nan
            continue
        base = prev if np.isfinite(prev) else price
        value = ((volume_period - volume) * base + volume * price) / volume_period
        out[idx] = value
        prev = value

    return out


def sample_series(length=96):
    prices = np.array(
        [100.0 + i * 0.23 + np.sin(i * 0.17) for i in range(length)],
        dtype=float,
    )
    volumes = np.array(
        [1000.0 + (i % 9) * 57.0 + i * 2.0 for i in range(length)],
        dtype=float,
    )
    return prices, volumes


def assert_series_close(actual, expected):
    actual = np.asarray(actual, dtype=float)
    expected = np.asarray(expected, dtype=float)
    assert actual.shape == expected.shape
    mask = ~(np.isnan(actual) & np.isnan(expected))
    assert np.allclose(actual[mask], expected[mask], atol=1e-10, rtol=0.0)


def test_evwma_matches_naive_absolute_mode():
    prices, volumes = sample_series()
    actual = ta_indicators.elastic_volume_weighted_moving_average(prices, volumes)
    expected = naive_evwma(prices, volumes)
    assert_series_close(actual, expected)


def test_evwma_matches_naive_volume_sum_mode():
    prices, volumes = sample_series()
    actual = ta_indicators.elastic_volume_weighted_moving_average(
        prices,
        volumes,
        length=7,
        absolute_volume_millions=200.0,
        use_volume_sum=True,
    )
    expected = naive_evwma(
        prices,
        volumes,
        length=7,
        absolute_volume_millions=200.0,
        use_volume_sum=True,
    )
    assert_series_close(actual, expected)


def test_evwma_stream_matches_batch():
    prices, volumes = sample_series(80)
    batch = ta_indicators.elastic_volume_weighted_moving_average(
        prices,
        volumes,
        length=9,
        absolute_volume_millions=220.0,
        use_volume_sum=True,
    )
    stream = ta_indicators.ElasticVolumeWeightedMovingAverageStream(
        length=9,
        absolute_volume_millions=220.0,
        use_volume_sum=True,
    )
    streamed = []
    for price, volume in zip(prices, volumes):
        value = stream.update(price, volume)
        streamed.append(value if value is not None else np.nan)
    streamed = np.asarray(streamed, dtype=float)
    stream.reset()
    streamed_after_reset = []
    for price, volume in zip(prices, volumes):
        value = stream.update(price, volume)
        streamed_after_reset.append(value if value is not None else np.nan)
    streamed_after_reset = np.asarray(streamed_after_reset, dtype=float)
    assert_series_close(streamed, batch)
    assert_series_close(streamed_after_reset, batch)


def test_evwma_batch_matches_single():
    prices, volumes = sample_series(84)
    batch = ta_indicators.elastic_volume_weighted_moving_average_batch(
        prices,
        volumes,
        length_range=(5, 9, 2),
        absolute_volume_millions=180.0,
        use_volume_sum=True,
    )
    assert batch["values"].shape == (3, len(prices))
    for row, length in enumerate(batch["lengths"]):
        single = ta_indicators.elastic_volume_weighted_moving_average(
            prices,
            volumes,
            length=int(length),
            absolute_volume_millions=180.0,
            use_volume_sum=True,
        )
        assert_series_close(batch["values"][row], single)


def test_evwma_errors():
    prices, volumes = sample_series(16)
    with pytest.raises(ValueError, match="input data slice is empty|Input data slice is empty"):
        ta_indicators.elastic_volume_weighted_moving_average(
            np.array([], dtype=float), np.array([], dtype=float)
        )
    with pytest.raises(ValueError, match="length mismatch"):
        ta_indicators.elastic_volume_weighted_moving_average(prices, volumes[:-1])
    with pytest.raises(ValueError, match="invalid length|Invalid length"):
        ta_indicators.elastic_volume_weighted_moving_average(prices, volumes, length=0)
    with pytest.raises(ValueError, match="invalid absolute_volume_millions|Invalid absolute_volume_millions"):
        ta_indicators.elastic_volume_weighted_moving_average(
            prices,
            volumes,
            absolute_volume_millions=0.0,
            use_volume_sum=False,
        )
